tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c

```diff
@@ -0,0 +1,650 @@
+		LOG_HEXDUMP_ERR(buf->data, buf->len, "HCI META EVT");
+	}
+}
+
+static void handle_ncp(struct net_buf *buf)
+{
+	struct bt_hci_evt_num_completed_packets *evt;
+	struct bt_hci_evt_hdr *hdr;
+	uint16_t handle, count;
+
+	hdr = net_buf_pull_mem(buf, sizeof(*hdr));
+
+	evt = (void *)buf->data;
+	handle = sys_le16_to_cpu(evt->h[0].handle);
+	count = sys_le16_to_cpu(evt->h[0].count);
+
+	LOG_DBG("sent %d packets", count);
+
+	while (count--) {
+		k_sem_give(&acl_pkts);
+	}
+}
+
+struct net_buf *alloc_l2cap_pdu(void);
+static void send_l2cap_packet(struct net_buf *buf, uint16_t cid);
+
+static void handle_att(struct net_buf *buf)
+{
+	uint8_t op = net_buf_pull_u8(buf);
+
+	switch (op) {
+	case BT_ATT_OP_NOTIFY:
+		LOG_INF("got ATT notification");
+		return;
+	case BT_ATT_OP_WRITE_RSP:
+		LOG_INF("got ATT write RSP");
+		SET_FLAG(flag_write_ack);
+		return;
+	case BT_ATT_OP_MTU_RSP:
+		LOG_INF("got ATT MTU RSP");
+		return;
+	default:
+		LOG_HEXDUMP_ERR(buf->data, buf->len, "payload");
+		TEST_FAIL("unhandled opcode %x\n", op);
+		return;
+	}
+}
+
+static void handle_l2cap(struct net_buf *buf)
+{
+	struct bt_l2cap_hdr *hdr;
+	uint16_t cid;
+
+	hdr = net_buf_pull_mem(buf, sizeof(*hdr));
+	cid = sys_le16_to_cpu(hdr->cid);
+
+	LOG_DBG("Packet for CID %u len %u", cid, buf->len);
+	LOG_HEXDUMP_DBG(buf->data, buf->len, "l2cap");
+
+	/* Make sure we don't have to recombine packets */
+	TEST_ASSERT(buf->len == hdr->len, "buflen = %d != hdrlen %d",
+	       buf->len, hdr->len);
+
+	TEST_ASSERT(cid == BT_L2CAP_CID_ATT, "We only support (U)ATT");
+
+	/* (U)ATT PDU */
+	handle_att(buf);
+}
+
+static void handle_acl(struct net_buf *buf)
+{
+	struct bt_hci_acl_hdr *hdr;
+	uint16_t len, handle;
+	uint8_t flags;
+
+	hdr = net_buf_pull_mem(buf, sizeof(*hdr));
+	len = sys_le16_to_cpu(hdr->len);
+	handle = sys_le16_to_cpu(hdr->handle);
+
+	flags = bt_acl_flags(handle);
+	handle = bt_acl_handle(handle);
+
+	TEST_ASSERT(flags == BT_ACL_START,
+	       "Fragmentation not supported");
+
+	LOG_DBG("ACL: conn %d len %d flags %d", handle, len, flags);
+	LOG_HEXDUMP_DBG(buf->data, buf->len, "HCI ACL");
+
+	handle_l2cap(buf);
+}
+
+static void recv(struct net_buf *buf)
+{
+	LOG_HEXDUMP_DBG(buf->data, buf->len, "HCI RX");
+
+	uint8_t code = buf->data[0];
+
+	if (bt_buf_get_type(buf) == BT_BUF_EVT) {
+		switch (code) {
+		case BT_HCI_EVT_CMD_COMPLETE:
+		case BT_HCI_EVT_CMD_STATUS:
+			handle_cmd_complete(buf);
+			break;
+		case BT_HCI_EVT_LE_META_EVENT:
+			handle_meta_event(buf);
+			break;
+		case BT_HCI_EVT_DISCONN_COMPLETE:
+			UNSET_FLAG(is_connected);
+			break;
+		case BT_HCI_EVT_NUM_COMPLETED_PACKETS:
+			handle_ncp(buf);
+			break;
+		default:
+			LOG_ERR("unhandled msg %x", code);
+			LOG_HEXDUMP_ERR(buf->data, buf->len, "HCI EVT");
+		}
+
+		/* handlers should take a ref if they want to access the buffer
+		 * later
+		 */
+		net_buf_unref(buf);
+		return;
+	}
+
+	if (bt_buf_get_type(buf) == BT_BUF_ACL_IN) {
+		handle_acl(buf);
```

Comment:
It looks like we are not doing anything with incoming data other than setting `flag_write_ack`. Is there something I've missed? If we are not using it, I would consider removing all the related code to simplify the test

Reply:
Useful for debugging (ie if the remote starts sending unexpected ATT messages). I'll keep it.

Reply:
Please remove `flag_write_ack` then